onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -1184,3 +1201,96 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider_CUDA, _In_ Or
   API_IMPL_END
 }
+
+ORT_API_STATUS_IMPL(OrtApis::CreateTensorRTProviderOptions, _Outptr_ OrtTensorRTProviderOptions** out) {
+  API_IMPL_BEGIN
+#ifdef USE_TENSORRT
+  *out = new OrtTensorRTProviderOptions();
+  (*out)->trt_int8_calibration_table_name = nullptr;
+  (*out)->trt_engine_cache_path = nullptr;
+  (*out)->trt_engine_decryption_lib_path = nullptr;
+  return nullptr;
+#else
+  ORT_UNUSED_PARAMETER(out);
+  return CreateStatus(ORT_FAIL, "TensorRT execution provider is not enabled in this build.");
+#endif
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::UpdateTensorRTProviderOptions,
+                    _Inout_ OrtTensorRTProviderOptions* tensorrt_provider_options,
+                    _In_reads_(num_keys) const char* const* provider_options_keys,
+                    _In_reads_(num_keys) const char* const* provider_options_values,
+                    size_t num_keys) {
+  API_IMPL_BEGIN
+#ifdef USE_TENSORRT
+  onnxruntime::ProviderOptions provider_options_map;
+  for (size_t i = 0; i != num_keys; ++i) {
+    if (provider_options_keys[i] == nullptr || provider_options_keys[i][0] == '\0' ||
+        provider_options_values[i] == nullptr || provider_options_values[i][0] == '\0') {
+      return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "key/value cannot be empty");
+    }
+
+    provider_options_map[provider_options_keys[i]] = provider_options_values[i];
+  }
+
+  onnxruntime::UpdateProviderInfo_Tensorrt(tensorrt_provider_options, reinterpret_cast<const onnxruntime::ProviderOptions&>(provider_options_map));
+  return nullptr;
+#else
+  ORT_UNUSED_PARAMETER(tensorrt_provider_options);
+  ORT_UNUSED_PARAMETER(provider_options_keys);
+  ORT_UNUSED_PARAMETER(provider_options_values);
+  ORT_UNUSED_PARAMETER(num_keys);
+  return CreateStatus(ORT_FAIL, "TensorRT execution provider is not enabled in this build.");
+#endif
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtApis::GetTensorRTProviderOptions, _Inout_ OrtAllocator* allocator,
+                    _Outptr_ char** ptr) {
+  API_IMPL_BEGIN
+#ifdef USE_TENSORRT
+  onnxruntime::ProviderOptions options = onnxruntime::GetProviderInfo_Tensorrt();
+  onnxruntime::ProviderOptions::iterator it = options.begin();
+  std::string options_str = "";
+
+  while (it != options.end()) {
+    if (options_str == "") {
+      options_str = it->first + "=" + it->second;
+    } else {
+      options_str += ";" + it->first + "=" + it->second;
+    }
+    it++;
+  }
+
+  *ptr = onnxruntime::StrDup(options_str, allocator);
+  return nullptr;
+#else
+  ORT_UNUSED_PARAMETER(allocator);
+  ORT_UNUSED_PARAMETER(ptr);
+  return CreateStatus(ORT_FAIL, "TensorRT execution provider is not enabled in this build.");
+#endif
+  API_IMPL_END
+}
+
+ORT_API(void, OrtApis::ReleaseTensorRTProviderOptions, _Frees_ptr_opt_ OrtTensorRTProviderOptions* ptr) {
+#ifdef USE_TENSORRT
+  if (ptr != nullptr) {
+    if (ptr->trt_int8_calibration_table_name != nullptr) {
+      delete ptr->trt_int8_calibration_table_name;
```

Comment:
Since a struct with the same name is defined in the C header file, the user can allocate the struct and its members in his/her code (without making a call to CreateTensorRTProviderOptions). If the user allocated ptr is passed to this function we can't know if the strings inside this struct were allocated using new or malloc.

Reply:
I think we can add a note to clearly tell users that please use this API only when OrtTensorRTProviderOptions instance is created by CreateTensorRTProviderOptions or they allocate memory using 'new' for the members in the struct. How does it sound? Or do you have better solution to address this problem with properly releasing memory.

Reply:
If we can properly document the usage of this API and people follow our API rule. I think we can avoid the repercussions as Pranav mentioned in the comment of CreateTensorRTProviderOptions.

Reply:
We should make our APIs easy to use correctly and difficult to use incorrectly. It's better to enforce these things in the code than through documentation. Besides ORT shouldn't be freeing/deleting memory that was allocated by the user in the first place; that's the user's responsibility. Otherwise we'll end up with a lot of github issues.

Reply:
I've made the modification.